File: local_agents/kit/plot/plot_iso36_rows_agg.py
#!/usr/bin/env python3
"""plot_iso36_rows_agg.py — the AGGREGATED variant of the row-format comparison
(mentor follow-up, 2026-08-25): SPEC collapsed to TWO bars (SPEC-int, SPEC-fp), plus a
Python comparison group, then the nine multilingual language groups. Appended beside
iso36_rows_* — never replacing it.

Columns per row:
  SPEC-int, SPEC-fp — ONE box each: the distribution of per-benchmark window-MEDIANS
      (14 int / 12 fp benchmarks; each benchmark votes once — pooling raw windows would
      let long benchmarks dominate).
  Python — scikit-learn, astropy, sympy from the matched-configuration SWE_iso8 replays
      (same cores 4-11 SMT-off, 100 ms windows; 8 shared groups). A fourth slot is
      reserved and marked: no fourth Python task has a banked matched capture, and no new
      profiling is run for it. The three fe_miss metrics (branch-direction MPKI, BTB
      MPKI, uop-cache MPKI) were never captured for Python — those cells carry a
      "to be measured" mark, deliberately not a zero.
  9 languages x 4 tasks — identical to iso36_rows_*.

    /home/thu/miniforge3/envs/infersuite-full/bin/python local_agents/kit/plot/plot_iso36_rows_agg.py
"""
from __future__ import annotations

import collections
import csv
import json
import os
import statistics as st
import sys

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt  # noqa: E402
import numpy as np  # noqa: E402

REPO = os.path.expanduser("~/InferSuite")
L3 = f"{REPO}/local_agents/ML_iso36/data/l3_study"
PYL3 = f"{REPO}/local_agents/SWE_iso8/data/l3_study"
OUT = f"{REPO}/local_agents/ML_iso36/plots"
SEL = f"{REPO}/local_agents/ML_typeid/selection_36_count.tsv"
sys.path.insert(0, f"{REPO}/spec26/kit/plot")
from spec_common import episodes as spec_episodes, windows as spec_windows  # noqa: E402

plt.rcParams.update({
    "font.family": "serif", "font.serif": ["DejaVu Serif"], "mathtext.fontset": "dejavuserif",
    "font.size": 10, "figure.dpi": 120, "savefig.dpi": 170, "savefig.bbox": "tight",
    "axes.grid": True, "grid.color": "#d8d8d8", "grid.linewidth": 0.5, "grid.alpha": 0.6,
    "axes.axisbelow": True,
})

LANGS = ["C", "C++", "Rust", "Go", "Java", "PHP", "Ruby", "JavaScript", "TypeScript"]
LCOL = {"C": "#0072B2", "C++": "#56B4E9", "Rust": "#D55E00", "Go": "#009E73",
        "Java": "#E69F00", "PHP": "#CC79A7", "Ruby": "#6b4fa0",
        "JavaScript": "#F0E442", "TypeScript": "#000000"}
PY_COL = "#8c510a"
INT_COL, FP_COL = "#4d4d4d", "#b3b3b3"
PYTASKS = ["scikit-learn", "astropy", "sympy"]

PANELS = [
    ("IPC", "IPC", "IPC (insn / cycle)", False),
    ("branch_MPKI", "brMPKI", "Branch MPKI (all mispredicts)", False),
    ("branchDir_MPKI", "branchDir_MPKI", "Branch-direction MPKI (cond.)", False),
    ("BTB_MPKI", "baclears_MPKI", "BTB MPKI (BAClears)", False),
    ("DSB_pct", "DSB_pct", "DSB coverage (% of delivered uops)", False),
    ("uopCache_MPKI", "dsb_miss_MPKI", "µop-cache (DSB) MPKI", False),
    ("codeRead_MPKI_L1I", "L1I_MPKI", "L1I MPKI (code-read)", False),
    ("L1D_MPKI", "L1D_MPKI", "L1D-load MPKI", False),
    ("L2_MPKI", "L2_MPKI", "L2-load MPKI", False),
    ("LLC_MPKI", "LLC_MPKI", "LLC MPKI", False),
    ("icache_data_stall_pct", "icache_data_stall_pct", "L1I stall (% cycles)", False),
    ("L1D_missrate_pct", "L1D_missrate_pct", "L1D miss rate (%)", False),
    ("L2_missrate_pct", "L2_missrate_pct", "L2-load miss rate (%)", False),
    ("LLC_missrate_pct", "LLC_missrate_pct", "LLC miss rate (%)", False),
    ("AMAT_cyc", "AMAT_cyc", "AMAT (cycles)", False),
    ("MLP", "MLP", "MLP (outstanding L1D misses)", False),
    ("dram_rd_GBs", "DRAM_read_GBs", "DRAM read bandwidth (GB/s)", False),
    ("ctx_per_cpu_s", "ctx_per_cpu_s", "Context switches (per CPU-s, log)", True),
]

sel = [r for r in csv.DictReader(open(SEL), delimiter="\t") if "__" in r.get("instance", "")]
by_lang = collections.defaultdict(list)
for r in sel:
    r["disp"] = r["instance"].split("__", 1)[1]
    by_lang[r["lang"]].append(r)

WIN = {}
for r in sel:
    p = f"{L3}/all_windows_{r['short']}.csv"
    if os.path.exists(p):
        WIN[r["short"]] = list(csv.DictReader(open(p)))
PYWIN = {t: (list(csv.DictReader(open(f"{PYL3}/all_windows_{t}.csv")))
             if os.path.exists(f"{PYL3}/all_windows_{t}.csv") else []) for t in PYTASKS}

print("loading SPEC per-window data (cached) ...")
SPEC = spec_episodes()
SWIN = {e["benchmark"]: spec_windows(e["dir"]) for e in SPEC}
SPEC_GROUPS = {"SPEC-int": [e["benchmark"] for e in SPEC if not e["fp"]],
               "SPEC-fp": [e["benchmark"] for e in SPEC if e["fp"]]}


def wvals(rows, metric, fence):
    return [float(x["value"]) for x in rows if x["metric"] == metric and x["fence"] == fence]


def spec_agg(gname, metric):
    """One value per benchmark: the median of its windows for this metric."""
    meds = []
    for b in SPEC_GROUPS[gname]:
        v = [r[metric] for r in SWIN[b] if metric in r and r[metric] is not None]
        if len(v) >= 5:
            meds.append(st.median(v))
    return meds


# ---- layout: SPEC-int | SPEC-fp || Python (3 + reserved) || 9 languages ----
SLOT, GGAP = 1.0, 2.2
cols, groups = [], []
x = 0.0
for g, col in (("SPEC-int", INT_COL), ("SPEC-fp", FP_COL)):
    cols.append((x, "specagg", g, col, f"{g} ({len(SPEC_GROUPS[g])})"))
    groups.append((g, x, x)); x += SLOT + GGAP * 0.35
x += GGAP * 0.65
x0 = x
for t in PYTASKS:
    cols.append((x, "python", t, PY_COL, t)); x += SLOT
cols.append((x, "reserved", "py4", PY_COL, "(4th: to fill)")); x += SLOT
groups.append(("Python", x0, x - SLOT)); x += GGAP
for lang in LANGS:
    x0 = x
    for r in by_lang[lang]:
        cols.append((x, "agent", r["short"], LCOL[lang], r["disp"])); x += SLOT
    groups.append((lang, x0, x - SLOT)); x += GGAP
XMAX = x - GGAP
PY_LO, PY_HI = groups[2][1], groups[2][2]

vals_dump = {"columns": [(k, key) for _x, k, key, _c, _t in cols],
             "spec_aggregation": "one value per benchmark = median of its windows; "
                                 "box over those per-benchmark medians"}
for fence in ("tool", "harness"):
    nrow = len(PANELS)
    fig, axes = plt.subplots(nrow, 1, figsize=(17, 2.1 * nrow), sharex=True)
    for ax, (ak, sk, label, use_log) in zip(axes, PANELS):
        data, pos, colors = [], [], []
        py_present = False
        for cx, kind, key, col, _t in cols:
            if kind == "specagg":
                v = spec_agg(key, sk)
            elif kind == "python":
                v = wvals(PYWIN[key], ak, fence)
            elif kind == "reserved":
                continue
            else:
                v = wvals(WIN.get(key, []), ak, fence)
            if len(v) < 3:
                continue
            if kind == "python":
                py_present = True
            data.append(v); pos.append(cx); colors.append(col)
            vals_dump[f"{fence}|{ak}|{key}"] = {
                "n": len(v), "median": st.median(v),
                "p5": float(np.percentile(v, 5)), "p95": float(np.percentile(v, 95))}
        if not data:
            ax.axis("off"); continue
        bp = ax.boxplot(data, positions=pos, widths=SLOT * 0.72, whis=(5, 95),
                        patch_artist=True, showfliers=False,
                        medianprops=dict(color="#111111", lw=1.2),
                        whiskerprops=dict(color="#777777", lw=0.7),
                        capprops=dict(color="#777777", lw=0.7))
        for box, c in zip(bp["boxes"], colors):
            box.set_facecolor(c)
            box.set_alpha(0.85)
            box.set_edgecolor("#666666" if c == "#F0E442" else "white")
            box.set_linewidth(0.5)
        if not py_present:
            ax.text((PY_LO + PY_HI) / 2, 0.5, "to be\nmeasured", transform=ax.get_xaxis_transform(),
                    ha="center", va="center", fontsize=8.5, style="italic", color="#8c510a")
        if use_log:
            ax.set_yscale("log")
        ax.set_xlim(-0.9, XMAX + 0.9)
        ax.set_ylabel(label, fontsize=9.5)
        ax.tick_params(length=0, labelsize=8)
        for _g, lo, hi in groups[:-1]:
            ax.axvline(hi + (SLOT + GGAP) / 2, color="#e6e6e6", lw=0.6, zorder=0)
        ax.axvline(groups[1][2] + (SLOT + GGAP * 0.65) / 2 + SLOT / 2, color="#7a8a99",
                   lw=1.1, ls=(0, (4, 3)), zorder=1)
    axb = axes[-1]
    axb.set_xticks([c[0] for c in cols])
    axb.set_xticklabels([c[4] for c in cols], rotation=90, fontsize=7.2)
    # one combined header over the two single-column SPEC groups (separate headers collide)
    axes[0].text((groups[0][1] + groups[1][2]) / 2, 1.04, "SPEC CPU 2026",
                 transform=axes[0].get_xaxis_transform(), ha="center", va="bottom",
                 fontsize=10.5, fontweight="bold", color=INT_COL)
    for g, lo, hi in groups[2:]:
        color = PY_COL if g == "Python" else LCOL[g]
        axes[0].text((lo + hi) / 2, 1.04, g, transform=axes[0].get_xaxis_transform(),
                     ha="center", va="bottom", fontsize=10.5, fontweight="bold", color=color)
    fig.suptitle(f"Per-window distributions, aggregated SPEC — {fence} fence · "
                 "SPEC-int / SPEC-fp = box over per-benchmark window-medians · Python = "
                 "matched-configuration replays · 100 ms windows, box = IQR, whiskers 5–95%",
                 fontsize=12.5, y=1.0)
    fig.tight_layout(rect=[0, 0, 1, 0.995])
    p = f"{OUT}/iso36_rows_agg_{fence}.png"
    fig.savefig(p)
    plt.close(fig)
    print(p)

json.dump(vals_dump, open(f"{OUT}/iso36_rows_agg_values.json", "w"), indent=1)
print(f"{OUT}/iso36_rows_agg_values.json")

File: local_agents/kit/plot/plot_iso36_rows_agg.R
#!/usr/bin/env Rscript
# plot_iso36_rows_agg.R -- house-style rendering of the aggregated per-window rows
# (mentor revision 2026-08-27): 16 metrics (MLP and AMAT dropped), one full-width row
# each; columns = SPEC-int | SPEC-fp (box over per-benchmark window-medians) | Python
# (3 tasks + reserved 4th) | 9 languages x 4 tasks. Violin + box (median line INSIDE the
# box, mean = white diamond). Outlier-heavy panels are AXIS-CAPPED, never data-trimmed:
# stats are computed on the full distributions and coord_cartesian cuts the view; a
# triangle marker + text names the off-scale maximum and its column.
#
# Inputs : local_agents/ML_iso36/data/l3_study/agg_rows_long.csv
# Outputs: local_agents/ML_iso36/plots/iso36_rows_agg_gg_{tool,harness}.{png,pdf}
#          local_agents/ML_iso36/plots/iso36_rows_agg_gg_numbers.csv
suppressPackageStartupMessages({
  library(ggplot2); library(hrbrthemes); library(dplyr); library(tidyr)
  library(yaml); library(ragg); library(scales); library(patchwork)
})
repo <- path.expand("~/InferSuite")
source(file.path(repo, "local_agents/kit/plot/iso36_style.R"))

d <- read.csv(file.path(repo, "local_agents/ML_iso36/data/l3_study/agg_rows_long.csv"),
              stringsAsFactors = FALSE)

langs <- c("C", "C++", "Rust", "Go", "Java", "PHP", "Ruby", "JavaScript", "TypeScript")
lcol <- c("SPEC-int" = "#4d4d4d", "SPEC-fp" = "#b3b3b3", "Python" = "#8c510a",
          "C" = "#0072B2", "C++" = "#56B4E9", "Rust" = "#D55E00", "Go" = "#009E73",
          "Java" = "#E69F00", "PHP" = "#CC79A7", "Ruby" = "#6b4fa0",
          "JavaScript" = "#F0E442", "TypeScript" = "#111111")
metrics <- c("IPC", "Branch MPKI", "Branch-direction MPKI", "BTB MPKI (BAClears)",
             "DSB coverage (%)", "uop-cache (DSB) MPKI", "L1I MPKI (code-read)",
             "L1D-load MPKI", "L2-load MPKI", "LLC MPKI", "L1I stall (% cycles)",
             "L1D miss rate (%)", "L2-load miss rate (%)", "LLC miss rate (%)",
             "DRAM read (GB/s)", "Context switches (/CPU-s)")
logm <- "Context switches (/CPU-s)"

# column order with spacer levels between groups; Python keeps a reserved 4th slot
ord <- c("SPEC-int", "SPEC-fp", ".s1",
         "scikit-learn", "astropy", "sympy", "(4th: to fill)", ".s2")
grp_of <- c("SPEC-int" = "SPEC-int", "SPEC-fp" = "SPEC-fp",
            "scikit-learn" = "Python", "astropy" = "Python", "sympy" = "Python",
            "(4th: to fill)" = "Python")
i <- 3
for (lg in langs) {
  cols <- unique(d$col[d$grp == lg])
  ord <- c(ord, cols, paste0(".s", i)); i <- i + 1
  for (cc in cols) grp_of[cc] <- lg
}
ord <- head(ord, -1)
d$col_f <- factor(d$col, levels = ord)
d$fillg <- grp_of[d$col]

num <- d |> group_by(fence, metric, grp, col) |>
  summarise(n = n(), median = median(value), mean = mean(value),
            p5 = quantile(value, .05), p95 = quantile(value, .95),
            max = max(value), .groups = "drop")
write.csv(num, file.path(repo, "local_agents/ML_iso36/plots/iso36_rows_agg_gg_numbers.csv"),
          row.names = FALSE)

groups_span <- function() {
  gs <- list()
  for (g in c("SPEC-int", "SPEC-fp", "Python", langs)) {
    idx <- which(ord %in% names(grp_of)[grp_of == g])
    if (length(idx)) gs[[g]] <- c(min(idx), max(idx))
  }
  gs
}
GS <- groups_span()

panel <- function(dd, m, show_x, fence, hdr = FALSE) {
  dm <- dd |> filter(metric == m)
  use_log <- m %in% logm
  # axis cap: default 1.15x the largest column p97 (whiskers always fit; extreme tails cut
  # from VIEW only, stats remain full-data). For the metrics where a WHOLE COLUMN is the
  # outlier (mentor 2026-08-27: BTB, the load-MPKI ladder), key the cap off the 90th
  # percentile of the columns' p95s instead, so the outlier columns cannot set the axis --
  # they go off-scale and are named by the red-triangle note.
  TRIM <- c("BTB MPKI (BAClears)", "L1D-load MPKI", "L2-load MPKI", "LLC MPKI")
  p97 <- dm |> group_by(col_f) |> summarise(q97 = quantile(value, .97),
                                            q95 = quantile(value, .95), .groups = "drop")
  cap <- if (m %in% TRIM) quantile(p97$q95, .90) * 1.2 else max(p97$q97) * 1.15
  mx <- dm |> group_by(col_f, fillg) |> summarise(mx = max(value), .groups = "drop") |>
    filter(mx > cap)
  p <- ggplot(dm, aes(x = col_f, y = value, fill = fillg)) +
    geom_violin(scale = "width", width = 0.85, linewidth = 0.15, colour = "grey40",
                adjust = 1.2, trim = TRUE, alpha = 0.75) +
    geom_boxplot(width = 0.28, outlier.shape = NA, linewidth = 0.25,
                 colour = "grey15", fill = "white", alpha = 0.9,
                 coef = 0) +
    stat_summary(fun = median, geom = "crossbar", width = 0.28, linewidth = 0.3,
                 colour = "black") +
    stat_summary(fun = mean, geom = "point", shape = 23, size = 1.1,
                 fill = "white", colour = "black", stroke = 0.35) +
    scale_fill_manual(values = lcol, guide = "none") +
    scale_x_discrete(drop = FALSE, labels = function(x) ifelse(grepl("^\\.s", x), "", x)) +
    labs(y = m, x = NULL) +
    theme_house(base_size = 8, axis_title_size = 8) +
    theme(axis.title.y = element_text(size = 6.6),
          axis.text.y = element_text(size = 5.8),
          panel.grid.major.x = element_blank(),
          plot.margin = margin(1, 4, 1, 2))
  if (use_log) {
    p <- p + scale_y_log10(labels = label_number(drop0trailing = TRUE))
  } else {
    p <- p + coord_cartesian(ylim = c(0, cap), clip = if (hdr) "off" else "on")
    if (nrow(mx) > 0) {
      mx <- mx |> arrange(desc(mx))
      lab3 <- paste(sprintf("%s %.3g", head(mx$col_f, 3), head(mx$mx, 3)), collapse = " · ")
      if (nrow(mx) > 3) lab3 <- paste0(lab3, sprintf("  (+%d more)", nrow(mx) - 3))
      p <- p +
        geom_point(data = mx, aes(x = col_f, y = cap * 0.985), shape = 17, size = 0.9,
                   colour = "#b2182b", inherit.aes = FALSE) +
        annotate("text", x = length(ord) - 0.5, y = cap * 0.93, hjust = 1, vjust = 1,
                 size = 1.7, colour = "#b2182b",
                 label = paste0("axis capped; off-scale max: ", lab3))
    }
  }
  if (show_x) {
    p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 5.2))
  } else {
    p <- p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
  }
  p
}

for (fence in c("tool", "harness")) {
  dd <- d |> filter(fence == !!fence)
  ps <- lapply(seq_along(metrics), function(k)
    panel(dd, metrics[k], show_x = (k == length(metrics)), fence, hdr = (k == 1)))
  # group headers on the first panel (SPEC-int and SPEC-fp are one column each —
  # separate headers collide, so they share one combined header)
  hdr <- ps[[1]] + annotate("text", x = 1.5, y = Inf, vjust = -0.55,
                            label = "SPEC CPU 2026", size = 2.0, fontface = "bold",
                            colour = "#4d4d4d")
  for (g in setdiff(names(GS), c("SPEC-int", "SPEC-fp"))) {
    hcol <- if (g == "TypeScript") "#111111" else lcol[[g]]
    hdr <- hdr + annotate("text", x = mean(GS[[g]]), y = Inf, vjust = -0.55,
                          label = g, size = 2.0, fontface = "bold", colour = hcol)
  }
  ps[[1]] <- hdr + theme(plot.margin = margin(9, 4, 1, 2))
  fig <- wrap_plots(ps, ncol = 1) + plot_annotation(
    title = sprintf("Per-window distributions, aggregated SPEC — %s fence", fence),
    subtitle = paste("SPEC-int / SPEC-fp = violin over per-benchmark window-medians (one vote per benchmark)",
                     "· violin + box, black bar = median, white diamond = mean",
                     "· red triangle = off-scale outlier (axis capped at 1.15x the largest column p97; stats use full data)"),
    caption = "Python: matched-configuration replays; the fe_miss trio and some rates are not banked for Python (empty slots) · 100 ms windows",
    theme = theme(plot.title = element_text(size = 11, face = "bold"),
                  plot.subtitle = element_text(size = 6.4, colour = "grey35"),
                  plot.caption = element_text(size = 5.4, colour = "grey45")))
  house_save(fig, file.path(repo, sprintf("local_agents/ML_iso36/plots/iso36_rows_agg_gg_%s", fence)),
             width = 14, height = 22)
}
